system/src/system_power_manager.cpp

```diff
@@ -552,52 +557,188 @@ void PowerManager::handleStateChange(battery_state_t from, battery_state_t to, b
+    } else {
+      chargingDebounceCount_ = 0;
+    }
+    chargedFaultCount_ = 0;
+    notChargingDebounceCount_ = 0;
+    lastChargedTimeStamp_ = 0;
+    battMonitorPeriod_ = BATTERY_STATE_CHANGE_CHECK_PERIOD;
+    DBG_PWR("chargingDebounceCount_: %d", chargingDebounceCount_);
+    if (chargingDebounceCount_ >= BATTERY_CHARGING_DEBOUNCE_COUNT) {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_CHARGING);
+    }
+  } else if (targetState == BATTERY_STATE_CHARGED) {
+    if (count) {
+      chargedFaultCount_++;
+    } else {
+      chargedFaultCount_ = 0;
+    }
+    notChargingDebounceCount_ = 0;
+    chargingDebounceCount_ = 0;
+    battMonitorPeriod_ = BATTERY_STATE_CHANGE_CHECK_PERIOD;
+    DBG_PWR("chargedFaultCount_: %d", chargedFaultCount_);
+    if (chargedFaultCount_ == 1) {
+      DBG_PWR("Set fault recharge threshold: 300mV");
+      PMIC power(true);
+      power.setRechargeThreshold(PMIC_FAULT_RECHARGE_THRESHOLD); // 300mV
+    } else if (chargedFaultCount_ >= BATTERY_CHARGED_FAULT_COUNT) {
+      DBG_PWR("Charged fault, disconnected");
+      // If we keep getting the charged event in a debouncing window, instead we assume that the battery is disconnected.
+      // No need to restore the normal threshold
+      confirmBatteryState(g_batteryState, BATTERY_STATE_DISCONNECTED);
     }
   }
-  return to;
 }
 
-void PowerManager::handlePossibleFaultLoop() {
-  if (faultSecondaryCounter_ == 1 && (millis() - possibleFaultTimestamp_ > DEFAULT_FAULT_WINDOW)) {
-      PMIC power;
-      power.setRechargeThreshold(PMIC_NORMAL_RECHARGE_THRESHOLD);
-      faultSecondaryCounter_ = 0;
-      faultSuppressed_ = millis();
+void PowerManager::clearIntermadiateBatteryState() {
+  chargingDebounceCount_ = 0;
+  vcellDebounceCount_ = 0;
+  chargedFaultCount_ = 0;
+  notChargingDebounceCount_ = 0;
+  lastChargedTimeStamp_ = 0;
+  battMonitorPeriod_ = BATTERY_STATE_NORMAL_CHECK_PERIOD;
+}
+
+void PowerManager::deduceBatteryStateLoop() {
+  // We don't need to deduce the battery state when the battery is used as the main power supplier
+  if (g_batteryState == BATTERY_STATE_DISCHARGING && battMonitorPeriod_ == BATTERY_STATE_NORMAL_CHECK_PERIOD) {
+    return;
+  }
+
+  // This should be executed only when charging is enabled and we got a charged event
+  if (!(config_.flags & HAL_POWER_CHARGE_STATE_DISABLE) && lastChargedTimeStamp_ != 0 && millis() - lastChargedTimeStamp_ >= BATTERY_CHARGED_MUTE_WINDOW) {
+    // We've passed the jitter window and the battery is charged indeed
+    DBG_PWR("Charged!");
+    confirmBatteryState(g_batteryState, BATTERY_STATE_CHARGED);
+    // Restore normal recharge threshold
+    DBG_PWR("Rstore normal threshold: 100mV");
+    PMIC power(true);
+    power.setRechargeThreshold(PMIC_NORMAL_RECHARGE_THRESHOLD); // 100mV
+    return;
+  }
+
+  if (millis() - batMonitorTimeStamp_ >= battMonitorPeriod_) {
+    batMonitorTimeStamp_ = millis();
+    if (config_.flags & HAL_POWER_CHARGE_STATE_DISABLE) {
```

Comment:
It's not guaranteed that the actual PMIC charging state is going to be what we intend it to be. We should watch out for that.